### PR TITLE
PR#1 for merging vectorized-read branch into master

### PR DESCRIPTION
**Pull request by samarthjain**
This PR introduces iterators for iterating through Parquet row groups and pages in a vectorized fashion. It also adds vectorized Parquet value decoders.

### CODE REVIEW

**Comment by samarthjain** [comment]
FYI - @rdblue, @prodeezy, @danielcweeks, @anjalinorwood

**Comment by prodeezy** [comment]
nice work @samarthjain, will take a look.

**Comment by rdblue** [comment]
Failures look like missing classes:
```
package org.apache.iceberg.parquet.arrow does not exist
```

File: parquet/src/main/java/org/apache/iceberg/parquet/ParquetUtil.java [hunk 1]
```diff
@@ -105,10 +112,14 @@ public static Metrics footerMetrics(ParquetMetadata metadata, MetricsConfig metr
             Types.NestedField field = fileSchema.findField(fieldId);
             if (field != null && stats.hasNonNullValue() && shouldStoreBounds(path, fileSchema)) {
               Literal<?> min = ParquetConversions.fromParquetPrimitive(
-                  field.type(), column.getPrimitiveType(), stats.genericGetMin());
+                  field.type(),
+                  column.getPrimitiveType(),
+                  stats.genericGetMin());
```
Comment (rdblue) [inline parquet/src/main/java/org/apache/iceberg/parquet/ParquetUtil.java]: Looks like an unrelated change?

File: parquet/src/main/java/org/apache/iceberg/parquet/ParquetUtil.java [hunk 2]
```diff
@@ -127,18 +138,16 @@ public static Metrics footerMetrics(ParquetMetadata metadata, MetricsConfig metr
...
+   * @return a list of offsets in ascending order determined by the starting position of the row groups
    */
   public static List<Long> getSplitOffsets(ParquetMetadata md) {
     List<Long> splitOffsets = new ArrayList<>(md.getBlocks().size());
     for (BlockMetaData blockMetaData : md.getBlocks()) {
       splitOffsets.add(blockMetaData.getStartingPos());
     }
     Collections.sort(splitOffsets);
-    return splitOffsets;
+    return ImmutableList.copyOf(splitOffsets);
```
Comment (rdblue) [inline parquet/src/main/java/org/apache/iceberg/parquet/ParquetUtil.java]: Same here, I don't think we need this?

File: parquet/src/main/java/org/apache/iceberg/parquet/ParquetUtil.java [hunk 3]
```diff
@@ -220,9 +231,132 @@ private static void increment(Map<Integer, Long> columns, int fieldId, long amou
...
+
+  public static boolean isVarWidthType(ColumnDescriptor desc) {
+    PrimitiveType primitive = desc.getPrimitiveType();
+    OriginalType originalType = primitive.getOriginalType();
+    if (originalType != null &&
+        originalType != OriginalType.DECIMAL &&
+        (originalType == OriginalType.ENUM ||
+            originalType == OriginalType.JSON ||
+            originalType == OriginalType.UTF8 ||
+            originalType == OriginalType.BSON)) {
```
Comment (rdblue) [inline parquet/src/main/java/org/apache/iceberg/parquet/ParquetUtil.java]: Since `OriginalType` is an enum, it would be better to use a `switch` statement for these.

**Comment by prodeezy** [comment]
Some of the classes from `./parquet/src/main/java/org/apache/iceberg/parquet/arrow` package weren't added in this PR e.g. `ArrowSchemaUtil` and `IcebergArrowColumnVector`  among others.

**Comment by samarthjain** [comment]
Closing this PR. I pushed a few commits to fix tests and remove unnecessary diffs. This PR is stale now. Will open new ones.
